applications/spectrum_analyzer/spectrum_analyzer.c

```diff
@@ -0,0 +1,475 @@
+#include <furi.h>
+#include <furi_hal.h>
+#include <gui/gui.h>
+#include <input/input.h>
+#include <stdlib.h>
+#include "spectrum_analyzer.h"
+
+void spectrum_analyzer_draw_scale(Canvas* canvas, const SpectrumAnalyzerModel* model) {
+
+    #define FREQ_BOTTOM_Y 50
+    #define FREQ_START_X 14
+    #define FREQ_LENGTH_X 102
+
+    // Draw line
+    canvas_draw_line(canvas, FREQ_START_X, FREQ_BOTTOM_Y, FREQ_START_X + FREQ_LENGTH_X, FREQ_BOTTOM_Y);
+    // Draw minor scale
+    for(int i = FREQ_START_X; i < FREQ_START_X + FREQ_LENGTH_X ; i += 5) {
+        canvas_draw_line(canvas, i, FREQ_BOTTOM_Y, i, FREQ_BOTTOM_Y + 2);
+    }
+    // Draw major scale
+    for(int i = FREQ_START_X; i < FREQ_START_X + FREQ_LENGTH_X ; i += 25) {
+        canvas_draw_line(canvas, i, FREQ_BOTTOM_Y, i, FREQ_BOTTOM_Y + 4);
+    }
+
+    // Draw scale tags
+    uint16_t tag_left;
+    uint16_t tag_center;
+    uint16_t tag_right;
+    char temp_str[18];
+
+    tag_center = user_freq;
+
+    switch (width) {
+        case NARROW:
+            tag_left = user_freq - 2;
+            tag_right = user_freq + 2;
+            break;
+        case ULTRAWIDE:
+            tag_left = user_freq - 40;
+            tag_right = user_freq + 40;
+            break;
+        default:
+            tag_left = user_freq - 10;
+            tag_right = user_freq + 10;
+    }
+
+    canvas_set_font(canvas, FontSecondary);
+    snprintf(temp_str, 18, "%u", tag_left);
+    canvas_draw_str_aligned(canvas, FREQ_START_X, 63, AlignCenter, AlignBottom, temp_str);
+    snprintf(temp_str, 18, "%u", tag_center);
+    canvas_draw_str_aligned(canvas, 128/2, 63, AlignCenter, AlignBottom, temp_str);
+    snprintf(temp_str, 18, "%u", tag_right);
+    canvas_draw_str_aligned(canvas, FREQ_START_X + FREQ_LENGTH_X - 1, 63, AlignCenter, AlignBottom, temp_str);
+}
+
+static void spectrum_analyzer_render_callback(Canvas* const canvas, void* ctx) {
+    const SpectrumAnalyzerModel* model = acquire_mutex((ValueMutex*)ctx, 25);
```

Comment:
Please switch to osMutex, we planning to drop value mutex in future releases

Reply:
Switched to osMutex